### ISSUE TITLE
SDK v10: Occasional error: java.lang.IllegalArgumentException: Flowable<ByteBuffer> emitted x bytes instead of the expected y bytes

### ISSUE THREAD

**Issue by rocketraman** [Azure/azure-storage-java#344]
### Which service(blob, file, queue, table) does this issue concern?
Blob

### Which version of the SDK was used?
10.0.1-Preview

### What problem was encountered?
Seemingly randomly (maybe every 1/20 uses), when doing an upload to an Azure blob store, I get the error `java.lang.IllegalArgumentException: Flowable<ByteBuffer> emitted x bytes instead of the expected y bytes` (where `x` < `y`) from the SDK, which comes from `FlowableUtil` in the rest client runtime (https://github.com/Azure/autorest-clientruntime-for-java/blob/v2/client-runtime/src/main/java/com/microsoft/rest/v2/util/FlowableUtil.java#L97).

Now, I'm doing some pretty bleeding edge stuff by using this preview SDK and integrating it's RX layer with Kotlin coroutines. However, the integration code is quite simple and shouldn't be causing any major issues. Therefore, I thought I'd log this issue here, because it may very well be a bug in the SDK.

### Have you found a mitigation/solution?

No

**Comment by rickle-msft** [comment]
Hi, @rocketraman. Thanks for raising this issue! This exception is thrown when the total number of bytes emitted by the Flowable does not match the blob size passed as a parameter. The user has to pass the size of the data because we can't tell from the Flowable itself how many bytes it will eventually emit, and we have to set a Content-Length header upfront. We throw this exception in the case of mismatch because we assume there was some calculation error and the user may get undesired behavior if their data is smaller or larger than what they specified. In particular, if we do not have this validation, it is possible the client will hang waiting for more bytes if the Flowable emits fewer bytes than specified. If the Flowable emits more bytes than specified, it is possible the user is losing data by not transferring it.


I hope this answers your question. Please double check the parameter value for the data size and the number of bytes you are actually emitting. If they match and you are still getting this Exception, please let us know. If you have any other concerns or questions, please do not hesitate to ask.

**Comment by rocketraman** [comment]
@rickle-msft I'll add some logging that lets me know how many bytes are being emitted.

**Comment by rocketraman** [comment]
@rickle-msft I think this is a bug in the SDK after all.

I replicated the conditions for this: it happens when the SDK has been used at least once, but then is idle for some time. The Flowable I pass to the SDK is being cancelled by the SDK before it has completely been read, and here is the stack trace which I see when the `doOnCancel` callback on the Flowable is called:

```
"nioEventLoopGroup-1-1@6352" daemon prio=5 tid=0x36 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
at com.rocketraman.azure.AzureStorageBlobStore$sendContent$flowable$3.run(AzureStorageBlobStore.kt:65)
	  at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.cancel(FlowableDoOnLifecycle.java:112)
	  at io.reactivex.internal.subscribers.BasicFuseableSubscriber.cancel(BasicFuseableSubscriber.java:158)
	  at io.reactivex.internal.subscribers.BasicFuseableSubscriber.cancel(BasicFuseableSubscriber.java:158)
	  at io.reactivex.internal.subscribers.BasicFuseableSubscriber.cancel(BasicFuseableSubscriber.java:158)
	  at io.reactivex.internal.subscribers.BasicFuseableSubscriber.cancel(BasicFuseableSubscriber.java:158)
	  at com.microsoft.rest.v2.http.NettyClient$AcquisitionListener$RequestSubscriber.operationComplete(NettyClient.java:410)
	  at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511)
	  at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485)
	  at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424)
	  at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:121)
	  at io.netty.util.internal.PromiseNotificationUtil.tryFailure(PromiseNotificationUtil.java:64)
	  at io.netty.channel.DelegatingChannelPromiseNotifier.operationComplete(DelegatingChannelPromiseNotifier.java:57)
	  at io.netty.channel.DelegatingChannelPromiseNotifier.operationComplete(DelegatingChannelPromiseNotifier.java:31)
	  at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511)
	  at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:504)
	  at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:483)
	  at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424)
	  at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:121)
	  at io.netty.util.internal.PromiseNotificationUtil.tryFailure(PromiseNotificationUtil.java:64)
	  at io.netty.channel.ChannelOutboundBuffer.safeFail(ChannelOutboundBuffer.java:702)
	  at io.netty.channel.ChannelOutboundBuffer.remove0(ChannelOutboundBuffer.java:294)
	  at io.netty.channel.ChannelOutboundBuffer.failFlushed(ChannelOutboundBuffer.java:639)
	  at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:744)
	  at io.netty.channel.AbstractChannel$AbstractUnsafe.flush0(AbstractChannel.java:945)
	  at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.flush0(AbstractNioChannel.java:360)
	  at io.netty.channel.AbstractChannel$AbstractUnsafe.flush(AbstractChannel.java:901)
	  at io.netty.channel.DefaultChannelPipeline$HeadContext.flush(DefaultChannelPipeline.java:1396)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:776)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:768)
	  at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:749)
	  at io.netty.handler.ssl.SslHandler.forceFlush(SslHandler.java:1776)
	  at io.netty.handler.ssl.SslHandler.wrapAndFlush(SslHandler.java:775)
	  at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:752)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:776)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:768)
	  at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:749)
	  at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:115)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:776)
	  at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:768)
	  at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:749)
	  at io.netty.channel.DefaultChannelPipeline.flush(DefaultChannelPipeline.java:1013)
	  at io.netty.channel.AbstractChannel.flush(AbstractChannel.java:248)
	  at com.microsoft.rest.v2.http.NettyClient$AcquisitionListener$RequestSubscriber.lambda$onNext$0(NettyClient.java:373)
	  at com.microsoft.rest.v2.http.NettyClient$AcquisitionListener$RequestSubscriber$$Lambda$71.240402964.run(Unknown Source:-1)
	  at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:163)
	  at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:-1)
	  at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	  at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:465)
	  at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	  at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	  at java.lang.Thread.run(Thread.java:748)
```

The key is this line here:

```
	  at io.netty.channel.ChannelOutboundBuffer.failFlushed(ChannelOutboundBuffer.java:639)
```

Attaching a debugger to see the `cause` at that line, it is because Netty has received a `java.io.IOException: Connection reset by peer` when doing the client socket write. So what is likely happening is that the Azure backend has timed out the idle connection from the SDK, and the SDK is not detecting and/or recovering from that properly.

I confirmed this via `netstat` on my Linux box... after the SDK has been used once, but before the connection has timed out:

```
tcp6       0      0 192.168.1.6:51850       40.86.232.206:443       ESTABLISHED 55661/java          
```

and then with the SDK idle, a few minutes later that connection is gone:

```
# netstat -anp | grep 40.86.232.206
<no output>
```

**Comment by rocketraman** [comment]
Some more information... I added logging on emitting, on the Flowable cancellation, and on Flowable complete, and here is what I see:

```
2018-07-27 11:19:47,819 INFO  --- [ioEventLoopGroup-1-8] STDOUT                                            : Flowable subscribed: PublisherCoroutine{Active}@1cf93e00.
2018-07-27 11:19:47,827 INFO  --- [ioEventLoopGroup-1-1] STDOUT                                            : Flowable cancelled.
2018-07-27 11:19:47,829 INFO  --- [ioEventLoopGroup-1-1] STDOUT                                            : Flowable complete, expected=692700, received=69496.
2018-07-27 11:19:47,829 INFO  --- [ioEventLoopGroup-1-1] STDOUT                                            : Flowable cancelled.
...
2018-07-27 11:19:52,359 INFO  --- [commonPool-worker-18] STDOUT                                            : TOTAL Emitted: 623204
2018-07-27 11:19:52,359 INFO  --- [commonPool-worker-18] STDOUT                                            : Flowable complete, expected=692700, received=692700.
2018-07-27 11:19:52,359 INFO  --- [commonPool-worker-18] STDOUT                                            : Flowable cancelled.
2018-07-27 11:19:52,361 INFO  --- [oEventLoopGroup-1-10] STDOUT                                            : Flowable cancelled.
2018-07-27 11:19:52,367 WARN  --- [     pool-2-thread-4] com.roc.azu.AzureStorageBlobStore                 : Request resulted in exception. Request=[] [[EXCEPTION: java.lang.IllegalArgumentException: Flowable<ByteBuffer> emitted 623204 bytes instead of the expected 692700 bytes.
	at com.microsoft.rest.v2.util.FlowableUtil$1.lambda$call$1(FlowableUtil.java:97) ~[client-runtime-2.0.0-beta3.jar:?]
	at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onComplete(FlowableDoOnEach.java:128) ~[rxjava-2.1.12.jar:?]
	at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onComplete(FlowableDoOnEach.java:135) ~[rxjava-2.1.12.jar:?]
	at io.reactivex.internal.subscribers.BasicFuseableSubscriber.onComplete(BasicFuseableSubscriber.java:120) ~[rxjava-2.1.12.jar:?]
	at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onComplete(FlowableDoOnEach.java:135) ~[rxjava-2.1.12.jar:?]
	at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.onComplete(FlowableDoOnLifecycle.java:94) ~[rxjava-2.1.12.jar:?]
	at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onComplete(FlowableDoOnEach.java:135) ~[rxjava-2.1.12.jar:?]
	at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.onComplete(FlowableDoOnLifecycle.java:94) ~[rxjava-2.1.12.jar:?]
	... suppressed 19 lines
]]
```

So, I believe what is happening is that the SDK is consuming the `Flowable`. That fails because of the connection reset described above, after 69496 bytes have been consumed:

```
2018-07-27 11:19:47,829 INFO  --- [ioEventLoopGroup-1-1] STDOUT                                            : Flowable complete, expected=692700, received=69496.
```

Then the SDK then executes a "retry" which assumes (wrongly in this case) that it can just re-subscribe to the same Flowable again, and it will read the input again from the beginning. However, because in this case the input Flowable is connected directly to a different socket, the retry is only able to read the *remaining* 623204 bytes, which is why we now see:

```
2018-07-27 11:19:52,132 INFO  --- [oEventLoopGroup-1-11] STDOUT                                            : Flowable subscribed: PublisherCoroutine{Active}@2f38f3db.
2018-07-27 11:19:52,359 INFO  --- [commonPool-worker-18] STDOUT                                            : TOTAL Emitted: 623204
2018-07-27 11:19:52,367 WARN  --- [     pool-2-thread-4] com.roc.azu.AzureStorageBlobStore                 : Request resulted in exception. Request=[] [[EXCEPTION: java.lang.IllegalArgumentException: Flowable<ByteBuffer> emitted 623204 bytes instead of the expected 692700 bytes.
```

I don't think the SDK should assume it can re-read the Flowable from the beginning for a retry -- I think it should probably keep the bytes it has read but not yet sent on the socket, and then after reestablishing the connection, just continue reading from the same Flowable as before. Alternatively, *this particular situation*, though not the underlying issue, could be fixed by checking for a valid connection to the server before starting to read the Flowable.

**Comment by rickle-msft** [comment]
Thank you for your thorough investigation. 

I think the key here is: "Then the SDK then executes a "retry" which assumes (wrongly in this case) that it can just re-subscribe to the same Flowable again, and it will read the input again from the beginning." You are indeed correct that the SDK assumes this, and it is a rather core assumption to the SDK. 

It is the philosophy of the SDK that retries are an essential and indispensable part of successfully interacting with the Storage service because of the way it helps to prevent data loss, and we want to give our customers the best chance of success possible. 

We do recognize the value of the scenario of uploading to Storage from another network stream, however, which we hope to address with the following additions to this library:

1. Enforce if possible or at least strongly indicate to the user that the Flowable they pass to us must be "replayable" that is, each new subscription to it will yield the same sequence of data. We are in conversation with the RxJava community about the best way to do this, but if you have suggestions, we would love to consider them! We are thinking perhaps an annotation in the form of @Nullable but instead @Replayble, or perhaps simply improving the error message you got to indicate to the user that this error may come from using a non-replayable Flowable and that will cause problems with retries.

2. Offer an uploadFromStream api that will take a Flowable which does not support this "replayability". This will accept a Flowable along with a number of buffers we are permitted to allocate and a bufferSize. We will handle chunking this input streams into these buffers and upload them (in parallel if we are allotted multiple buffers) using retries to ensure there is no data loss in uploading this stream.

We believe that these two features combined will give customers the best path forward. FYI, there was a whole [conversation](https://github.com/Azure/azure-storage-blob-go/issues/31)
 about this on our Go SDK that you can also reference for more background.

In the meantime, if this is blocking you, you may set the tryCount on the RequestRetryOptions in PipelineOptions to be 1, which will ensure only the initial request is sent and no retries are made. We strongly discourage this because of the aforementioned value of retires, but it is possible.


Please let us know if you have any concerns or suggestions about this path forward!

**Comment by RikkiGibson** [comment]
Hi @rocketraman. It's expected that a `Flowable<ByteBuffer>` will provide an equivalent sequence of ByteBuffers each time you subscribe to it, or if it's not possible, will emit an error on subsequent subscriptions, which is what we do in the case of [HttpResponse.body()](https://github.com/Azure/autorest-clientruntime-for-java/blob/v2/client-runtime/src/main/java/com/microsoft/rest/v2/http/NettyClient.java#L725).

Depending on the nature of the data source for the request, it's usually not too difficult to make it repeatable using an operator like [Flowable.defer](http://reactivex.io/RxJava/javadoc/io/reactivex/Flowable.html#defer-java.util.concurrent.Callable-) or [Flowable.generate](http://reactivex.io/RxJava/javadoc/io/reactivex/Flowable.html#generate-java.util.concurrent.Callable-io.reactivex.functions.BiConsumer-io.reactivex.functions.Consumer-).

You can also apply [Flowable.cache](http://reactivex.io/RxJava/javadoc/io/reactivex/Flowable.html#cache--) to your request body if you'd like to make your Flowable repeatable by caching the content in memory until everything has been successfully uploaded.

**Comment by rocketraman** [comment]
Thank you both for your comments. I agree it is not the job of the SDK to take care of buffering, and also that it is quite reasonable that the `Flowable<ByteBuffer>` be replayable, at least until the alternative `uploadFromStream` api can be added. I do believe there are some improvements that can come out of this:

1. This deserves some documentation
2. I agree with @rickle-msft 's suggestion that the error message I encountered could be more explicit -- it would have saved me a lot of time.
3. The particular situation I encountered of the Azure backend connection timing out, should not require a retry to be successful. The connection timeout should be detected by an event (`channel.closeFuture()`?) and the client state reset such that the SDK knows to reestablish the connection again before the next transfer. This will reduce latency of these requests.

**Comment by rocketraman** [comment]
> I think the error message: Flowable<ByteBuffer> emitted 623204 bytes instead of the expected 692700 bytes. could reasonably be augmented with something like:

If you know the context is a retry you can add the additional wording conditionally. Something like:

```
Flowable<ByteBuffer> emitted 623204 bytes on retry, instead of the expected 692700 bytes. Note that automatic retry support requires the input Flowable to support replay.
```

and the original error message for the non-retry case.

> Currently when we pull a connection out of the pool, we do a "health check", before attempting to use it to send a request. I'll plan on spending some time with a longer running stress test to see if there's a more reliable way for us to avoid attempting to send a request through a closed channel.

I just confirmed with a breakpoint that Netty isn't returning the expected result from `isActive()`, at least on my system -- even though the connection is gone (confirmed with netstat), Netty still thinks it is active and writable.

I fired up Wireshark to check out what is happening at the TCP level. I see that on my environment Azure (or perhaps another intermediate system) seems to close the connection with an abrupt RST rather than the usual FIN / FIN-ACK dance, so it looks like the connection is in a half-closed state. I don't think any hardware on my end is doing this, so it is possible this is an Azure storage problem. I have saved the packet capture -- let me know if you want to see it.

I don't know if Netty can detect a half-closed connection state without actually sending some bytes on the wire as is being done now. Some possible solutions: 

1) Fix the server to do a proper connection close instead of an abrust RST.

2) Use a Netty IdleStateHandler to close the connection properly from the client, *before* it gets half-closed by the server. 

3) Use a periodic ping/pong type no-op message to keep the connection open.

> Thank you very much for reporting the problem, and I think it's really cool that you're using Kotlin coroutines with this SDK. I feel like that's the most compelling language/environment to take advantage of the async features.

Yes, the integration was pretty simple with Kotlin coroutines rx extensions. It is all working very nicely.

**Comment by RikkiGibson** [comment]
This is really useful information. Just poking around a bit, I wonder if netty/netty#3193 is related.

**Comment by rocketraman** [comment]
> I wonder if netty/netty#3193 is related.

I don't think so. I believe this was a case of a Netty server sending a RST instead of FIN, and I think the OP eventually said the RST wasn't the issue at all.

**Comment by RikkiGibson** [comment]
Got it.

> 1. Fix the server to do a proper connection close instead of an abrust RST.

@rickle-msft might know the right people to direct this question to.

> 2. Use a Netty IdleStateHandler to close the connection properly from the client, before it gets half-closed by the server.

Is it necessary to do this before the server sends the RST? If something like setting a shorter connection keep-alive mitigates the problem, that would be great, but I guess it seems a little bit brittle.

> 3. Use a periodic ping/pong type no-op message to keep the connection open.

Definitely possible, but I wonder if this should be combined with some kind of higher-level keep-alive mechanism so we don't attempt to keep a connection open (for example) for hours or days.

**Comment by rocketraman** [comment]
@seguler I have not put it into production yet, but yes, I am successfully using the new Storage SDK in a test environment. All the issues that would block me from going into production have been resolved and/or answered.

**Comment by rickle-msft** [comment]
@rocketraman Quick update for you. We are planning to put out a release candidate mid next week which will have some improved documentation and error reporting per our earlier discussion. We will also add a method to take a non replayable Flowable to the TransferManager shortly after GA. Once the RC goes out, I'll close this issue as we are tracking the addition of this Transfermanager feature in another issue.

**Comment by RikkiGibson** [comment]
Cool, let's just make sure we're tracking the issue about sending data into half-closed connections.

**Comment by rickle-msft** [comment]
I am going to close this issue as it is being tracked [here](https://github.com/Azure/autorest-clientruntime-for-java/issues/467), which is the more appropriate place for it. Please feel free to reopen this issue if there are further concerns directly related to this library.
